File: cpp/Glacier2/session/SessionManager.cpp
// Copyright (c) ZeroC, Inc.

#include "SessionManager.h"
#include "DefaultPokeSession.h"

#include <iostream>

using namespace Server;
using namespace std;

SessionManager::SessionManager(Ice::ObjectAdapterPtr adapter) : _sessionAdapter(std::move(adapter)) {}

std::optional<Glacier2::SessionPrx>
SessionManager::create(
    std::string userId,
    std::optional<Glacier2::SessionControlPrx> sessionControl,
    const Ice::Current& current)
{
    // sessionControl is not null because we configured Glacier2.Server.Endpoints in the Glacier2 router
    // configuration file.
    assert(sessionControl);

    // Create a new session servant and add it to the adapter with a UUID identity. The name component of the
    // identity is the session token.
    Ice::ObjectPrx proxy = _sessionAdapter->addWithUUID(
        make_shared<DefaultPokeSession>(_sessionAdapter, *sessionControl, shared_from_this()));

    string sessionToken = proxy.ice_getIdentity().name;
    _tokenToUserId[sessionToken] = userId;

    cout << "Created session #" << sessionToken << " for user '" << userId << "'" << endl;
    return Ice::uncheckedCast<Glacier2::SessionPrx>(proxy);
}

std::optional<std::string>
SessionManager::getUserId(const std::string& token)
{
    const auto& it = _tokenToUserId.find(token);
    return it == _tokenToUserId.end() ? std::nullopt : std::optional<std::string>{it->second};
}

void
SessionManager::removeToken(const std::string& token)
{
    _tokenToUserId.erase(token);
}

File: swift/Glacier2/Session/Sources/Server/SessionManager.swift
// Copyright (c) ZeroC, Inc.

import Foundation
import Glacier2
import Ice

/// SessionManager is an Ice servant that implements Slice interface Glacier2::SessionManager. It creates sessions
/// and resolves user IDs from session tokens.
/// Note: This demo implementation is not thread-safe. A real implementation should support concurrent calls.
class SessionManager: Glacier2.SessionManager, UserIdResolver {
    private let adapter: Ice.ObjectAdapter
    private var tokenToUserId: [String: String] = [:]

    /// Constructs a SessionManager servant.
    /// - Parameter adapter: The object adapter that hosts the session servants.
    init(adapter: Ice.ObjectAdapter) {
        self.adapter = adapter
    }

    func create(
        userId: String,
        control: Glacier2.SessionControlPrx?,
        current: Ice.Current
    ) async throws -> Glacier2.SessionPrx? {
        // control is not nil because we configured Glacier2.Server.Endpoints in the Glacier2 router
        // configuration file.
        guard let sessionControl = control else {
            fatalError("SessionControl is nil")
        }

        // Create a new session servant and add it to the adapter with a UUID identity. The name component of the
        // identity is the session token.
        let sessionServant = DefaultPokeSession(
            adapter: adapter,
            sessionControl: sessionControl,
            userIdResolver: self
        )

        let proxy = try adapter.addWithUUID(servant: Ice.Disp(sessionServant)) as CatchThemAll.PokeSessionPrx

        let sessionToken = proxy.ice_getIdentity().name
        tokenToUserId[sessionToken] = userId

        print("Created session #\(sessionToken) for user '\(userId)'")
        return proxy
    }

    func getUserId(_ token: String) -> String? {
        return tokenToUserId[token]
    }

    func removeToken(_ token: String) {
        tokenToUserId.removeValue(forKey: token)
    }
}
